hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+    return resourcesAvailable;
+  }
+
+  @VisibleForTesting
+  protected void scheduleContainer(Container container) {
+    if (maxOppQueueLength <= 0) {
+      startAllocatedContainer(container);
+      return;
+    }
+    if (queuedGuaranteedContainers.isEmpty() &&
+        queuedOpportunisticContainers.isEmpty() &&
+        this.utilizationManager.hasResourcesAvailable(container)) {
+      startAllocatedContainer(container);
+    } else {
+      LOG.info("No available resources for container {} to start its execution "
+          + "immediately.", container.getContainerId());
+      boolean isQueued = true;
+      if (container.getContainerTokenIdentifier().getExecutionType() ==
+          ExecutionType.GUARANTEED) {
+        queuedGuaranteedContainers.put(container.getContainerId(), container);
+        // Kill running opportunistic containers to make space for
+        // guaranteed container.
+        killOpportunisticContainers(container);
+      } else {
+        if (queuedOpportunisticContainers.size() <= maxOppQueueLength) {
+          LOG.info("Opportunistic container {} will be queued at the NM.",
+              container.getContainerId());
+          queuedOpportunisticContainers.put(
+              container.getContainerId(), container);
+        } else {
+          isQueued = false;
+          LOG.info("Opportunistic container [{}] will not be queued at the NM" +
+              "since max queue length [{}] has been reached",
+              container.getContainerId(), maxOppQueueLength);
+          container.sendKillEvent(
+              ContainerExitStatus.KILLED_BY_CONTAINER_SCHEDULER,
+              "Opportunistic container queue is full.");
+        }
+      }
+      if (isQueued) {
+        try {
+          this.context.getNMStateStore().storeContainerQueued(
+              container.getContainerId());
+        } catch (IOException e) {
+          LOG.warn("Could not store container state into store..", e);
+        }
+      }
+    }
+  }
+
+  private void killOpportunisticContainers(Container container) {
+    List<Container> extraOpportContainersToKill =
+        pickOpportunisticContainersToKill(container.getContainerId());
+    // Kill the opportunistic containers that were chosen.
+    for (Container contToKill : extraOpportContainersToKill) {
+      contToKill.sendKillEvent(
+          ContainerExitStatus.KILLED_BY_CONTAINER_SCHEDULER,
+          "Container Killed to make room for Guaranteed Container.");
+      oppContainersMarkedForKill.put(contToKill.getContainerId(), contToKill);
+      LOG.info(
+          "Opportunistic container {} will be killed in order to start the "
+              + "execution of guaranteed container {}.",
+          contToKill.getContainerId(), container.getContainerId());
+    }
+  }
+
+  private void startAllocatedContainer(Container container) {
+    LOG.info("Starting container [" + container.getContainerId()+ "]");
+    scheduledToRunContainers.put(container.getContainerId(), container);
```

Comment:
I agree it is not technically needed. But, being a scheduler, I felt it would come in handy to have a data-structure of running containers.
One other use-case I forsee is if the container is updated (if we implement container promotion, for instance), we would have to move it from one list to another, which is cumbersome.
I vote to keep it as it is for the time-being and move it if required.